File: src/actions/link.js
"use strict";

const path = require("path");
const chalk = require("chalk");
const flatry = require("flatry");
const de = require("directory-exists");
const symlink = require("symlink-dir");
const { fmt, exit, paths, lstats, ispath_abs } = require("../utils/toolbox.js");

module.exports = async (args) => {
	let { registrypath } = paths;
	let { path: p } = args;

	if (p) if (!ispath_abs(p)) p = path.resolve(p);

	let cwd = p || process.cwd();
	let dirname = path.basename(cwd);
	let pkgpath = `${registrypath}/${dirname}`;

	let [err, res] = await flatry(de(cwd));
	if (err || !res) process.exit();

	// If package exists error.
	[err, res] = await flatry(de(pkgpath));
	if (err) process.exit();
	if (res) {
		[, res] = await flatry(lstats(pkgpath));
		let type = res.is.symlink ? "Symlink " : "";
		let msg = `${type}?/ exists. Remove it and try again.`;
		exit([fmt(msg, chalk.bold(dirname))]);
	}

	await symlink(cwd, pkgpath);
};

File: src/parser/nim/actions/link.nim
import os, asyncdispatch, json, strformat

import ../utils/[chalk, paths]

proc nlcli_link*(s: string = "{}") {.async.} =
    let registrypath = paths["registrypath"]

    let jdata = parseJSON(s)
    let path = jdata{"path"}.getStr()

    let p = (
        if path.len != 0 and not isAbsolute(path):
            absolutePath(path)
        else: ""
    )

    let cwd = if p.len != 0: p else: parentDir(currentSourcePath())
    let dirname = splitPath(cwd).tail
    let destination = fmt"{registrypath}/{dirname}"

    # If folder exists give error.
    if not dirExists(cwd): quit() # Confirm cwd exists.

    # If folder exists give error.
    if dirExists(destination) or symlinkExists(destination):
        # Check if folder is a symlink.
        let `type`= if symlinkExists(destination): "Symlink " else: ""
        quit(`type` & dirname.chalk("bold") & "/ exists. Remove it and try again.")

    createSymlink(cwd, destination) # Create symlink.
